Linkta-org/core
# 91: Set up unit testing environments

## Description



Fixes #91 

### 1. Adds unit testing environments for the client/ and server/ directories using [Vitest](https://vitest.dev/). 
- the client/ and server/ directories each have their own tests/ directory that will contain all tests
- each directory has a `test` and `test:watch` npm script for running tests once and in watch mode respectively
- Config details:
  - client:
    - configured within the existing `vite.config.js` file
    - test environment is jsdom 
  - server: 
    - configured within `vitest.config.ts` 
    - test environment is node
  - all files matching `**/*.unit.test.[jt]s?(x)` will be run by Vitest

### 2. Deletes old testing placeholders from root directory


## Type of change



- [X] Other 🗒️Testing

## How Can this be tested? Testing Plan to review this PR



1. navigate to the client/ directory
2. run `npm run test` - you should see one test, 'true is true', passing
3. navigate to the server/ directory
4. run `npm run test` - again, you should see one test, 'true is true', passing


## Checklist:




## Additional Information:



A few notes on the selection of the testing framework -

I’ve chosen to use Vitest for unit (and soon integration) testing in both the client and server apps. This is in large part because Vitest has native support for working with ES Modules. Since Linkta explicitly uses `"type": "module"` for both applications, choosing a testing solution that gets along well with ESM was a high priority. The options I considered were Jest, Vitest, and Mocha (supported by other libraries). 

The biggest issue with testing ESM seems to be mocking. ES Modules are immutable, so as [this guide](https://spencerfeng.medium.com/mastering-the-art-of-mocking-es-modules-in-jest-7a615ac0e879) on mocking ESM with Jest explains, ‘the exports of ES modules are frozen and cannot be altered during runtime.’ This can make using Jest with ES Modules [a little complicated](https://jestjs.io/docs/ecmascript-modules).

While mocking ESM with Jest is certainly possible, their docs note that its support for ESM is built on experimental Node APIs and is subject to change. Mocha [recommends](https://mochajs.org/#nodejs-native-esm-support) using the mocking library [testdouble](https://github.com/testdouble/testdouble.js) for ESM. Testdouble looks great and well documented, but may lack the community support of more popular options like Jest and Vitest. 

In the interest of choosing a framework that both meets our needs and is as accessible for the team as possible, I think Vitest is the best choice ⚡
